dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataExport.java

```diff
@@ -7,272 +7,101 @@
-        System.out.println("partial export: metadataexport -i handle -f filename");
-        System.exit(exitCode);
-    }
-
-    /**
-     * main method to run the metadata exporter
-     *
-     * @param argv the command line arguments given
-     * @throws Exception if error occurs
-     */
-    public static void main(String[] argv) throws Exception {
-        // Create an options object and populate it
-        CommandLineParser parser = new PosixParser();
-
+    private Options constructOptions() {
         Options options = new Options();
 
         options.addOption("i", "id", true, "ID or handle of thing to export (item, collection, or community)");
+        options.getOption("i").setType(String.class);
         options.addOption("f", "file", true, "destination where you want file written");
+        options.getOption("f").setType(OutputStream.class);
+        options.getOption("f").setRequired(true);
         options.addOption("a", "all", false,
                           "include all metadata fields that are not normally changed (e.g. provenance)");
+        options.getOption("a").setType(boolean.class);
         options.addOption("h", "help", false, "help");
+        options.getOption("h").setType(boolean.class);
 
-        CommandLine line = null;
-
-        try {
-            line = parser.parse(options, argv);
-        } catch (ParseException pe) {
-            System.err.println("Error with commands.");
-            printHelp(options, 1);
-            System.exit(0);
-        }
 
-        if (line.hasOption('h')) {
-            printHelp(options, 0);
-        }
+        return options;
+    }
 
-        // Check a filename is given
-        if (!line.hasOption('f')) {
-            System.err.println("Required parameter -f missing!");
-            printHelp(options, 1);
+    public void internalRun() throws Exception {
+        if (help) {
+            handler.logInfo("\nfull export: metadata-export -f filename");
+            handler.logInfo("partial export: metadata-export -i handle -f filename");
+            printHelp();
+            return;
         }
-        String filename = line.getOptionValue('f');
 
-        // Create a context
-        Context c = new Context(Context.Mode.READ_ONLY);
-        c.turnOffAuthorisationSystem();
-
-        // The things we'll export
-        Iterator<Item> toExport = null;
-        MetadataExport exporter = null;
+        DSpaceCSV dSpaceCSV = metadataDSpaceCsvExportService
+            .handleExport(context, exportAllItems, exportAllMetadata, handle,
+                          handler);
+        handler.writeFilestream(context, filename, dSpaceCSV.getInputStream(), "exportCSV");
+        context.restoreAuthSystemState();
+        context.complete();
+    }
 
-        // Export everything?
-        boolean exportAll = line.hasOption('a');
+    public void setup() throws ParseException {
+        context = new Context();
+        context.turnOffAuthorisationSystem();
```

Comment:
Again, please resolve prior feedback here.

Reply:
This is a port from the CLI script that initially does this on the following places:
* https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataExport.java#L226
* https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataExport.java#L275

The reason for this being that IF the script is executed from the CLI you don't want AuthorizeExceptions. This script can now not only be executed from the CLI but also from the REST api (the same class is used for both of these). The REST api has a user (which right now, which is certainly an admin as scripts & processes can only be executed by admins), but the CLI doesn't.

So a possible solution would be, is adding an a mandatory user parameter for the CLI, that way we can leave this out.

What do you think ?

Reply:
Again, you've accidentally introduced a flaw here.  As you linked to previously, it's perfectly OK to turn off/on authorizations _within the same method_.  However, now that this is across separate methods, it will require finer tuning.  You may need to turn it off & on & off & on again only when needed within specific methods.  Or, maybe there's a way to turn it off in the calling code in order to surround two other internal method calls...then turn it back on after those calls.

But, again, this PR's current behavior is flawed as we don't allow for turning off authorizations across methods...that can have unintended consequences and accidentally bypass AuthorizeExceptions that we *expect to occur*.  

I'm OK with us requiring a user parameter on CLI if that helps us get around this problem.

Reply:
As above, I'd suggest moving `turnOffAuthorizationSystem()` and the subsequent `setCurrentUser()` from `setup()` to `internalRun()`.  I don't see why `setup()` needs a Context at all, so I say that it shouldn't have one.

Reply:
Agreed, we will make adjustments to keep the context object limited to methods & remove it as a variable in the class.